rfcs/text/0000-data-processing-robustness.md

```diff
@@ -0,0 +1,73 @@
+
+### User Stories
+
+- As a DCP developer, I don't want to make urgent fixes to software due to differences or errors in data, so that development efforts can be planned and managed in a controlled manner, rather than a reactive one.
+- As a DCP submitter, I want my data to be ingested promptly, without waiting for DCP developers to make modifications to handle my experiment, so that my lab can move on to other tasks.
+- As a DCP data consumer, I don't want to receive scientifically incorrect, inaccurate or incomplete data caused by developers attempting to work around problematic data, so that I can spend my time on actual research.
+
+## Detailed Design
+
+In this RFC, we require two main strategies for making changes across the DCP that will ensure greater reliability and predictability of data processing. These strategies will reduce coupling between components, providing the opportunity for development teams to react in a controlled, planned manner whenever data is encountered that fails to meet the assumptions or expectations of DCP software components.
+
+The proposed strategies are:
+1. *Log and Continue*
+2. *Repair and Recover*
+
+### Log and Continue
+
+- DCP components define their expectations of any data they receive (see 
+[Querying DSS by Metadata Schema Version(s)](https://github.com/HumanCellAtlas/dcp-community/blob/master/rfcs/text/0011-query-by-metadata-schema-versions.md)
+- DCP components ignore data they retrieve if it is mismatched against their expectations (e.g. bundles contain unrecognised file formats such as PDFs)
+- DCP components skip processing of data and fail gracefully when they encounter an error, rather than crashing or producing incorrect results.
+_ DCP components log all errors in a manner that allows them to be detected and addressed.  See [Monitoring for production systems](https://github.com/HumanCellAtlas/dcp-community/blob/master/rfcs/text/0002-monitoring-for-production-systems.md).
+- DCP components report correct error statuses (e.g. via HTTP response codes) to clients as appropriate.
+
+
+### Repair and Recover
+
+- DCP components include operator admin functions to manually trigger automated steps that may have failed
+- In case of errors that causes some data to be skipped, DCP components ensure that all data expected to be handled together (e.g. all data from a single project) complete before making some of the data available
+
+
+### Unresolved Questions
+
+- It is expected that this RFC to raise questions that require more detailed specification of the requirements.
+- How each component determines and reports compliance to these principles needs to be defined.
+
+### Prior work:
+- [The Harmful Consequences of the Robustness Principle](https://tools.ietf.org/html/draft-iab-protocol-maintenance-03) - Despite the name, this IETF draft very much supports a premise of this RFC, which is to not attempt process unexpected data, rather skip and move on in a principled manner.
```

Comment:
Your _skip and move on_ statement is more reflective of Postel's principle, which this internet draft is responding against - _Time and experience shows that negative consequences to interoperability accumulate over time if implementations apply the robustness principle._

The guidance includes:

[Virtuous Intolerance](https://tools.ietf.org/html/draft-iab-protocol-maintenance-03#section-7.2)
> A well-specified protocol includes rules for consistent handling of
   aberrant conditions.  This increases the changes that implementations
   have interoperable handling of unusual conditions.

   > Intolerance of any deviation from specification, where
   implementations generate fatal errors in response to observing
   undefined or unusual behaviour, can be harnessed to reduce occurrences
   of aberrant implementations.  Choosing to generate fatal errors for
   unspecified conditions instead of attempting error recovery can
   ensure that faults receive attention.

   > This improves feedback for new implementations in particular.  When a
   new implementation encounters an intolerant implementation, it
   receives strong feedback that allows problems to be discovered
   quickly.

Reply:
This is an interesting and nuanced discussion.  

Postel's principle says to ignore and move on, while this specification is about record error, move on, and fix later.  So superficially it might look like  Postel's, however, it is about both not stopping processing and not losing data or creating incorrect results.

This treats the DCP more like a credit card system than twitter.  Don't let one bad transaction go undetected but don't let it bring down the whole system either.

Right now, changes upstream are hard to make because of the fear of breaking downstream.  So it makes it very difficult to bring in new data.

Reply:
The example referenced from HTTP/2 in the section that I quoted above:

_An endpoint MAY immediately terminate an HTTP/2 connection that does not meet these TLS requirements with a connection error (Section 5.4.1) of type INADEQUATE_SECURITY._ which is not record, move on, and fix later.

Reply:
This is a very important conversation, I am glad to see it happening. I would imagine no one would want any component to go down due to data or metadata that is not well understood. So in this way, I am echoing resilience of the operation of components to the unknown is critical. It was my experience over years of working with many scientists that researchers will use any output given to them, even on a failed run with extensive logging indicating the data is wrong; the mere existence of output is validation enough that the analysis was successful. It is important that, while components do not fail, the system does not progress metadata or data that is not understood and deletes or does not store any derived product from meta(data) that encounters an error. Clear signals should exist in components throughout the life-cycle of the data to alert operations to the (meta)data so the scenario can be resolved (and that resolution hopefully automated to be used to handle the next occurrence) without breaking the operation of the component but with such an open system as ours, it is important we assure users can not access outputs associated with meta(data) in question.

Reply:
Exactly @TimothyTickle .  Outputting incorrect results on error is far worse crashing.  This RFC forbids such behavior..